### PR TITLE
Refactor: Use HtSP for storing plugins.

### PR DESCRIPTION
**Pull request by Rot127**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [ ] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

Refactor: Use `HtSP` for storing plugins.
This saves iterations when searching for a plugin of a given name. It also allows to save plugin configurations and data in hash table, all addressed by their name. Hence, usage is more streamlined.

**Test plan**

All green

**Closing issues**

...

### CODE REVIEW

**Comment by Rot127** [comment]
This one seems green now. The Woodpecker job failed for unrelated reasons. Can you restart it please? My authentication is not working, can't log in. No idea why.

**Comment by Rot127** [comment]
The Woodpecker jobs seems to have another Heisenbug?
```
[XX] 892 ms   db/archos/darwin-arm64/dbg maps as flags
RZ_NOPLUGINS=1 /private/var/folders/bc/yvvw2g2s4jlg9r8x7tf61xv800009d/T/woodpecker-local-2079229593/workspace/prefix/bin/rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -eflirt.sigdb.load.system=false -eflirt.sigdb.load.home=false -N -d -qc 'fl@F:maps~hello~[1-]
echo --
dm*~hello~[0-2]
' bins/mach0/hello-macos-arm64
-- stdout
--- expected
+++ actual
@@ -1,6 +1,8 @@
 16384 hello_macos_arm64.r_x
 16384 hello_macos_arm64.rw
 16384 hello_macos_arm64.rw.
+f0000
+
 16384 hello_macos_arm64.r
 --
```

File: librz/util/list.c [hunk 1]
```diff
@@ -296,6 +296,24 @@ RZ_API RZ_OWN RzList *rz_list_new_from_array(RZ_NONNULL const void **arr, size_t
 	return l;
 }
 
+/**
+ * \brief Allocates a new RzList and adds all elements of the iterator \p iter to it.
+ * \p iter keeps the ownership over the values.
+ *
+ * \return The produced list. Or NULL in case of failure.
+ **/
+RZ_API RZ_OWN RzList *rz_list_new_from_iterator(RZ_BORROW RZ_NONNULL RzIterator *iter) {
```
Comment (XVilka) [inline librz/util/list.c:308]: If NONNULL, then please add an assert

File: librz/main/rz-hash.c [hunk 2]
```diff
@@ -130,11 +130,17 @@ static void rz_hash_show_algorithms(RzHashContext *ctx) {
 
 	printf("flags  algorithm      license    author\n");
 
+	RzIterator *iter = ht_sp_as_iter(ctx->rh->plugins);
+	RzList *plugin_list = rz_list_new_from_iterator(iter);
```
Comment (XVilka) [inline librz/main/rz-hash.c:134]: Check for NULL

File: librz/main/rz-gg.c [hunk 3]
```diff
@@ -68,20 +69,26 @@ static int usage(int v) {
 }
 
 static void list(RzEgg *egg) {
-	RzListIter *iter;
-	RzEggPlugin *p;
 	printf("shellcodes:\n");
-	rz_list_foreach (egg->plugins, iter, p) {
+	RzIterator *iter = ht_sp_as_iter(egg->plugins);
+	RzList *plugin_list = rz_list_new_from_iterator(iter);
```
Comment (XVilka) [inline librz/main/rz-gg.c:74]: Check for NULL

File: librz/egg/egg.c [hunk 4]
```diff
@@ -548,14 +552,17 @@ RZ_API int rz_egg_shellcode(RzEgg *egg, const char *name) {
 			return true;
 		}
 	}
+	rz_iterator_free(iter);
 	return false;
 }
 
 RZ_API int rz_egg_encode(RzEgg *egg, const char *name) {
```
Comment (XVilka) [inline librz/egg/egg.c]: Since you added asserts, please add argument attributes too

File: librz/egg/egg.c [hunk 5]
```diff
@@ -532,10 +534,12 @@ RZ_API char *rz_egg_option_get(RzEgg *egg, const char *key) {
 }
 
 RZ_API int rz_egg_shellcode(RzEgg *egg, const char *name) {
```
Comment (XVilka) [inline librz/egg/egg.c]: Same

File: librz/crypto/crypto.c [hunk 6]
```diff
@@ -6,12 +6,12 @@
 #include <rz_util.h>
 #include "rz_crypto_plugins.h"
 
+static RzCryptoPlugin *crypto_static_plugins[] = { &rz_crypto_plugin_aes, &rz_crypto_plugin_aes_cbc, &rz_crypto_plugin_base64, &rz_crypto_plugin_base91, &rz_crypto_plugin_blowfish, &rz_crypto_plugin_cps2, &rz_crypto_plugin_des, &rz_crypto_plugin_punycode, &rz_crypto_plugin_rc2, &rz_crypto_plugin_rc4, &rz_crypto_plugin_rc6, &rz_crypto_plugin_rol, &rz_crypto_plugin_ror, &rz_crypto_plugin_rot, &rz_crypto_plugin_serpent, &rz_crypto_plugin_xor, &rz_crypto_plugin_sm4_ecb };
```
Comment (XVilka) [inline librz/crypto/crypto.c]: Split it into multiple lines please

File: librz/core/cio.c [hunk 7]
```diff
@@ -479,16 +479,24 @@ RZ_API RzCmdStatus rz_core_io_plugin_print(RzIOPlugin *plugin, RzCmdStateOutput
...
+
 	if (!io) {
 		return RZ_CMD_STATUS_ERROR;
 	}
 	rz_cmd_state_output_array_start(state);
 	rz_cmd_state_output_set_columnsf(state, "sssss", "perm", "license", "name", "uri", "description");
-	rz_list_foreach (io->plugins, iter, plugin) {
+
+	RzIterator *iter = ht_sp_as_iter(io->plugins);
+	RzList *plugin_list = rz_list_new_from_iterator(iter);
```
Comment (XVilka) [inline librz/core/cio.c:491]: Check for NULL

File: librz/core/cio.c [hunk 8]
```diff
@@ -479,16 +479,24 @@ RZ_API RzCmdStatus rz_core_io_plugin_print(RzIOPlugin *plugin, RzCmdStateOutput
  * \param state Specify how plugins shall be printed
  */
 RZ_API RzCmdStatus rz_core_io_plugins_print(RzIO *io, RzCmdStateOutput *state) {
```
Comment (XVilka) [inline librz/core/cio.c]: Add attributes then

File: librz/core/chash.c [hunk 9]
```diff
@@ -34,19 +34,22 @@ RZ_API RzCmdStatus rz_core_hash_plugin_print(RzCmdStateOutput *state, const RzHa
 RZ_API RzCmdStatus rz_core_hash_plugins_print(RzHash *hash, RzCmdStateOutput *state) {
```
Comment (XVilka) [inline librz/core/chash.c]: Same, attributes

File: librz/core/cdebug.c [hunk 10]
```diff
@@ -325,23 +325,30 @@ RZ_API RzCmdStatus rz_core_debug_plugins_print(RzCore *core, RzCmdStateOutput *s
...
-	RzListIter *iter;
-	RzDebugPlugin *plugin;
 	RzCmdStatus status;
 	if (!dbg) {
 		return RZ_CMD_STATUS_ERROR;
 	}
 	rz_cmd_state_output_array_start(state);
-	rz_list_foreach (dbg->plugins, iter, plugin) {
+	RzIterator *iter = ht_sp_as_iter(dbg->plugins);
+	RzList *plugin_list = rz_list_new_from_iterator(iter);
```
Comment (XVilka) [inline librz/core/cdebug.c:335]: Check for NULL

File: librz/core/ccrypto.c [hunk 11]
```diff
@@ -34,19 +35,27 @@ static RzCmdStatus core_crypto_plugin_print(RzCmdStateOutput *state, const RzCry
...
-	const RzCryptoPlugin *plugin = NULL;
 	RzCmdStatus status;
-	RzListIter *it;
 	rz_cmd_state_output_array_start(state);
 	if (state->mode == RZ_OUTPUT_MODE_STANDARD) {
 		rz_cons_println("algorithm      license    author");
 	}
-	rz_list_foreach (cry->plugins, it, plugin) {
+	RzIterator *iter = ht_sp_as_iter(cry->plugins);
+	RzList *plugin_list = rz_list_new_from_iterator(iter);
```
Comment (XVilka) [inline librz/core/ccrypto.c:44]: Check for nULL

File: librz/bp/bp_plugin.c [hunk 12]
```diff
@@ -45,24 +47,32 @@ RZ_API bool rz_bp_plugin_del(RzBreakpoint *bp, RZ_BORROW RZ_NONNULL RzBreakpoint
...
 	}
+	rz_iterator_free(iter);
 	return false;
 }
 
 // TODO: deprecate
 RZ_API void rz_bp_plugin_list(RzBreakpoint *bp) {
-	RzListIter *iter;
+	RzIterator *iter = ht_sp_as_iter(bp->plugins);
+	RzList *plugin_list = rz_list_new_from_iterator(iter);
```
Comment (XVilka) [inline librz/bp/bp_plugin.c:67]: Check for NULL

Comment (XVilka) [reply librz/bp/bp_plugin.c:67]: Maybe rename the function into `_print()` while you are on it? For consistency with other plugins, also remove the comment about deprecation?

**Review by XVilka — CHANGES_REQUESTED** [review]
Quite nice changes!

File: librz/core/casm.c [hunk 13]
```diff
@@ -162,11 +168,14 @@ RZ_API RzCmdStatus rz_core_asm_plugin_print(RzCore *core, RzAsmPlugin *ap, const
 RZ_API RzCmdStatus rz_core_asm_plugins_print(RzCore *core, const char *arch, RzCmdStateOutput *state) {
 	int i;
 	RzAsm *a = core->rasm;
+	RzIterator *iter = ht_sp_as_iter(a->plugins);
+	RzList *plugin_list = rz_list_new_from_iterator(iter);
```
Comment (XVilka) [inline librz/core/casm.c:173]: You forgot to check for NULL here as well.

File: librz/core/cbin.c [hunk 14]
```diff
@@ -4902,24 +4903,40 @@ RZ_IPI RzCmdStatus rz_core_binxtr_plugin_print(const RzBinXtrPlugin *bx, RzCmdSt
 RZ_API RzCmdStatus rz_core_bin_plugins_print(RzBin *bin, RzCmdStateOutput *state) {
 	rz_return_val_if_fail(bin && state, RZ_CMD_STATUS_ERROR);
 
+	RzCmdStatus status;
+	rz_cmd_state_output_array_start(state);
+
+	RzIterator *iter = ht_sp_as_iter(bin->plugins);
+	RzList *plugin_list = rz_list_new_from_iterator(iter);
```
Comment (XVilka) [inline librz/core/cbin.c:4910]: And here

File: test/unit/test_list.c [hunk 15]
```diff
@@ -218,6 +219,30 @@ bool test_rz_list_sort5(void) {
...
+bool test_rz_list_from_iter(void) {
+	HtUP *alpha_ht = ht_up_new(NULL, NULL);
+	char *unordered_alphabeth[] = { "b", "w", "k", "a", "c", "d", "e", "f", "g", "h", "i", "j", "l", "m", "n", "o", "p", "q", "r", "s", "t", "u", "v", "x", "y", "z" };
+	char *lower[] = { "a", "b", "c", "d", "e", "f", "g", "h", "i", "j", "k", "l", "m", "n", "o", "p", "q", "r", "s", "t", "u", "v", "w", "x", "y", "z" };
+
+	for (size_t i = 0; i < 26; i++) {
+		ht_up_insert(alpha_ht, i, (void *)unordered_alphabeth[i]);
+	}
+	RzIterator *iter = ht_up_as_iter(alpha_ht);
+	RzList *list = rz_list_new_from_iterator(iter);
```
Comment (XVilka) [inline test/unit/test_list.c:231]: `mu_assert_neq(list, NULL)`

**Review by XVilka — APPROVED** [review]
LGTM with a few small nitpicks

File: test/unit/test_list.c (hunk 15 above)
Comment (wargio) [reply test/unit/test_list.c:231]: there is a function for checking for null.

**Review by XVilka — APPROVED** [review]
@wargio take a look and lets merge this

**Review by wargio — APPROVED** [review]

**Pull request merged**